packages/worker/src/worker-tuner.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  FixedSizeSlotSupplier,
+  ResourceBasedTunerOptions,
+  WorkerTuner as NativeWorkerTuner,
+  SlotSupplier as NativeSlotSupplier,
+} from '@temporalio/core-bridge';
+import { Duration, msToNumber } from '@temporalio/common/lib/time';
+
+export type WorkerTuner = ResourceBasedTuner | TunerHolder;
+
+type ResourceBasedSlotsForType = ResourceBasedSlotOptions & {
+  tunerOptions: ResourceBasedTunerOptions;
+};
+
+/**
+ * @experimental
+ *
+ * Controls how slots are handed out for a specific task type.
+ *
+ * For now, only {@link ResourceBasedSlotOptions} and {@link FixedSizeSlotSupplier} are supported,
+ * but we may add support for custom tuners in the future.
+ */
+export type SlotSupplier = ResourceBasedSlotsForType | FixedSizeSlotSupplier;
+
+/**
+ * @experimental
+ *
+ * Options for a specific slot type within a {@link ResourceBasedSlotsForType}
+ */
+export interface ResourceBasedSlotOptions {
+  // Amount of slots that will be issued regardless of any other checks
+  minimumSlots: number;
+  // Maximum amount of slots permitted
+  maximumSlots: number;
+  // Minimum time we will wait (after passing the minimum slots number) between handing out new
+  // slots
+  rampThrottle: Duration;
+}
+
+/**
+ * @experimental
+ *
+ * This tuner attempts to maintain certain levels of resource usage when under load. You do not
+ * need more than one instance of this when using it for multiple slot types.
+ */
+export interface ResourceBasedTuner {
+  // Options for the tuner
+  tunerOptions: ResourceBasedTunerOptions;
+  // Options for workflow task slots. Defaults to a minimum of 2 slots and a maximum of 1000 slots
+  // with no ramp throttle
+  workflowTaskSlotOptions?: ResourceBasedSlotOptions;
+  // Options for activity task slots. Defaults to a minimum of 1 slots and a maximum of 2000 slots
+  // with 50ms ramp throttle
+  activityTaskSlotOptions?: ResourceBasedSlotOptions;
+  // Options for local activity task slots. Defaults to a minimum of 1 slots and a maximum of 2000
+  // slots with 50ms ramp throttle
+  localActivityTaskSlotOptions?: ResourceBasedSlotOptions;
+}
+
+/**
+ * @experimental
+ *
+ * This tuner allows for different slot suppliers for different slot types.
+ */
+export interface TunerHolder {
+  workflowTaskSlotSupplier: SlotSupplier;
+  activityTaskSlotSupplier: SlotSupplier;
+  localActivityTaskSlotSupplier: SlotSupplier;
+}
+
+export function asNativeTuner(tuner: WorkerTuner): NativeWorkerTuner {
```

Comment:
Does this need to be exported. I am thinking we want the smallest user-facing API surface we can get away with.

Reply:
It does so it can be used from the other parts of the package. It is not publicly exported at a top level.